### PR TITLE
DT-1232: 🎨 Add in service based mfa without code checking

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaServiceBasedController.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class MfaServiceBasedController(
+  private val jwtAuthenticationSuccessHandler: JwtAuthenticationSuccessHandler,
+) {
+
+  @GetMapping("/service-mfa-challenge")
+  fun mfaChallengeRequest(
+    @RequestParam user_oauth_approval: String?,
+  ): ModelAndView = ModelAndView(
+    "serviceMfaChallenge",
+    mapOf("smokeCode" to "123456", "user_oauth_approval" to user_oauth_approval),
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaServiceBasedController.kt:30]: skeleton controller that just outputs smoke code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaServiceBasedController.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  @PostMapping("/service-mfa-challenge")
+  fun mfaChallenge(
+    request: HttpServletRequest,
+    response: HttpServletResponse,
+    authentication: Authentication,
+  ): String {
+
+    // TODO: Check token to ensure passed mfa
+    jwtAuthenticationSuccessHandler.updateMfaInRequest(request, response, authentication)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaServiceBasedController.kt:41]: skeleton service that automatically approves any token

File: src/main/resources/templates/serviceMfaChallenge.html [hunk 3]
```diff
@@ -0,0 +1,93 @@
+<!DOCTYPE html>
+<html xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout" xmlns="http://www.w3.org/1999/html" lang="en"
+      xmlns:th="http://www.thymeleaf.org"
+      layout:decorate="~{fragments/layout}">
+<head>
+  <title th:with="mfaPreference=${param.mfaPreference ?: mfaPreference}"
+         th:text="${#strings.equals(mfaPreference, 'TEXT')} ? 'Text Message Verification' : 'Email Verification'"></title>
```
Comment (petergphillips) [inline src/main/resources/templates/serviceMfaChallenge.html:7]: skeleton html file to get us going

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/integration/MfaServiceBasedClientLoginSpecification.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        assertThat(claims.getClaim("user_name")).isEqualTo("AUTH_CHANGE_TEST")
+      }
+  }
+
+  @Test
+  fun `Sign in as nomis user`() {
+    clientAccess(
+      {
+        loginPage.isAtPage().submitLogin("ITAG_USER", "password")
+        serviceBasedMfaFromAuthorizeEmailPage.isAtPage().submitCode()
```
Comment (petergphillips) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/integration/MfaServiceBasedClientLoginSpecification.kt:58]: need to add a few more test scenarios - will do in next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserContextApprovalHandler.kt [hunk 5]
```diff
@@ -46,16 +51,38 @@ class UserContextApprovalHandler(
...
+
+    if (!userDetails.passedMfa && clientNeedsMfa(authorizationRequest.clientId)) {
+      // found a client that requires mfa
+      userApprovalRequest["requireMfa"] = true
+    }
 
     return userApprovalRequest
   }
+
+  private fun clientNeedsMfa(clientId: String): Boolean {
```
Comment (JamieDraperUK) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/UserContextApprovalHandler.kt:83]: Cheeky bit of duplication here.

**Review by JamieDraperUK — APPROVED** [review]
LGTM

**Pull request merged**